### PR TITLE
ES6 refactor lib/node/NodeMainTemplatePlugin.js

### PR DESCRIPTION
**Pull request by ayastreb**
**What kind of change does this PR introduce?**

ES6 refactoring

**Did you add tests for your changes?**

No

**If relevant, link to documentation update:**

N/A

**Summary**
Convert codebase to ES6. #4099 

**Does this PR introduce a breaking change?**

No

**Other information**

My first commit to webpack 🤓

### CODE REVIEW

File: lib/node/NodeMainTemplatePlugin.js [hunk 1]
```diff
@@ -2,130 +2,160 @@
...
+					"if(installedChunks[chunkId])",
+					this.indent([
+						"return installedChunks[chunkId][2];"
+					]),
+					"// load the chunk and return promise to it",
+					"var promise = new Promise(function(resolve, reject) {",
+					this.indent([
+						"installedChunks[chunkId] = [resolve, reject];",
+						"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify("/" + chunkFilename), {
+							hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
```
Comment (ayastreb) [inline lib/node/NodeMainTemplatePlugin.js]: This part of [code](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L73-L91) is repeated two more times in this file - [here](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L118-L136) and [here](https://github.com/webpack/webpack/blob/master/lib/node/NodeMainTemplatePlugin.js#L155-L173).
Maybe I should try to refactor it out? Or is it not in the scope of ES6 refactoring?

File: lib/node/NodeMainTemplatePlugin.js [hunk 2]
```diff
@@ -2,130 +2,160 @@
...
+			}
+			return source;
+		});
+		mainTemplate.plugin("require-extensions", function(source, chunk) {
+			if(chunk.chunks.length > 0) {
+				return this.asString([
+					source,
+					"",
+					"// uncatched error handler for webpack runtime",
+					this.requireFn + ".oe = function(err) {",
```
Comment (sokra) [inline lib/node/NodeMainTemplatePlugin.js]: Could you try to refactor these string concatenations to template strings?

Comment (ayastreb) [inline lib/node/NodeMainTemplatePlugin.js]: I've thought about it, but then I found https://jsperf.com/es6-string-literals-vs-string-concatenation/14 and seems like string concatenation is faster than template literals, but template literal is faster than array.join (which is used in Template.asString).
Looking on the code in other modules I couldn't find the prefered way to deal with concatenation - in some places it's `+` in other places it's template literals.

Do you prefer template literals?

File: lib/node/NodeMainTemplatePlugin.js [hunk 3]
```diff
@@ -112,28 +112,28 @@ module.exports = class NodeMainTemplatePlugin {
 					"return installedChunks[chunkId][2] = promise;"
 				]);
 			} else {
-				const request = this.applyPluginsWaterfall("asset-path", JSON.stringify("./" + chunkFilename), {
-					hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
-					hashWithLength: (length) => "\" + " + this.renderCurrentHashCode(hash, length) + " + \"",
+				const request = this.applyPluginsWaterfall("asset-path", JSON.stringify(`./${chunkFilename}`), {
+					hash: `" + ${this.renderCurrentHashCode(hash)} + "`,
+					hashWithLength: (length) => `" + ${this.renderCurrentHashCode(hash, length)} + \"`,
```
Comment (sokra) [inline lib/node/NodeMainTemplatePlugin.js]: ``ode(hash, length)} + \"`,``
``ode(hash, length)} + "`,``

File: lib/node/NodeMainTemplatePlugin.js [hunk 4]
```diff
@@ -148,25 +148,25 @@ module.exports = class NodeMainTemplatePlugin {
 			const hotUpdateMainFilename = this.outputOptions.hotUpdateMainFilename;
 			const chunkMaps = chunk.getChunkMaps();
 			const currentHotUpdateChunkFilename = this.applyPluginsWaterfall("asset-path", JSON.stringify(hotUpdateChunkFilename), {
-				hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
-				hashWithLength: (length) => "\" + " + this.renderCurrentHashCode(hash, length) + " + \"",
+				hash: `" + ${this.renderCurrentHashCode(hash)} + "`,
+				hashWithLength: (length) => `" + ${this.renderCurrentHashCode(hash, length)} + \"`,
```
Comment (sokra) [inline lib/node/NodeMainTemplatePlugin.js]: ``rentHashCode(hash, length)} + \"`,``
``rentHashCode(hash, length)} + "`,``

File: lib/node/NodeMainTemplatePlugin.js [hunk 5]
```diff
@@ -69,21 +69,21 @@ module.exports = class NodeMainTemplatePlugin {
 					"var promise = new Promise(function(resolve, reject) {",
 					this.indent([
 						"installedChunks[chunkId] = [resolve, reject];",
-						"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify("/" + chunkFilename), {
-							hash: "\" + " + this.renderCurrentHashCode(hash) + " + \"",
-							hashWithLength: (length) => "\" + " + this.renderCurrentHashCode(hash, length) + " + \"",
+						"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify(`/${chunkFilename}`), {
+							hash: `" + ${this.renderCurrentHashCode(hash)} + "`,
+							hashWithLength: (length) => `" + ${this.renderCurrentHashCode(hash, length)} + \"`,
```
Comment (sokra) [inline lib/node/NodeMainTemplatePlugin.js]: ``rrentHashCode(hash, length)} + \"`,``
``rrentHashCode(hash, length)} + "`,``

**Comment by sokra** [comment]
Thanks

**Pull request merged**